### ISSUE TITLE
Fatal error: Western (ISO Latin 1) cannot encode <Unicode string>: file SwiftTaggerID3/String.swift, line 23

### ISSUE THREAD

**Issue by peternguyenxq** [NCrusher74/SwiftTaggerID3#18]
Hello @NCrusher74 
My name is Peter. 
I am working with @Coleman
And I am using SwiftTaggerID3 to updating ID3
I have issues with edit ID3 for files that contain Unicode Characters in Its ID3 tags: title, comment, composer, album

Here is my code: 

           let mp3File = try Mp3File(location: URL(fileURLWithPath: filePath))
            var tag = try mp3File.tag()

            tag.album = mp3Data.album ?? ""
            tag.artist = mp3Data.artist ?? ""
            tag.title = mp3Data.name ?? ""
            tag.composer = mp3Data.composer ?? ""
            tag.arranger = mp3Data.remixer ?? ""
            tag[comment: "Comment", .eng] = mp3Data.comment ?? ""
            
            try mp3File.write(
                tag: &tag,
                version: .v2_4,
                outputLocation: URL(fileURLWithPath: filePath))

And Here is the error code: Fatal error: Western (ISO Latin 1) cannot encode <Unicode string>: file SwiftTaggerID3/String.swift, line 23

Could you please help to handle Unicode Characters?

Thanks and Bests
Peter

**Comment by NCrusher74** [comment]
Hi Peter!

Without having the actual data you're working with, I'm only guessing here, but I made some changes to how string encoding was handled, and hopefully it will solve the problem.

Try using the `encoded-fixes` branch and see if that solves your issues. If it does, I'll go ahead and merge it. If not, I may need to ask my mentor for help, because I'm not all that confident that I'm using the right approach for how determining the string encoding is handled.

**Comment by NCrusher74** [comment]
You should also be aware that the changes I've implemented in the `encoding-fixes` branch attempt to initialize a `String.Encoding` instance from a given `String`, and then translate that into an encoding byte. I can see three potential issues with this approach.

1) While `utf16` and `utf16BigEndian` are permissible in the ID3 specs, a lot of third party apps won't recognize those encoding options. `Yate` and `MediaInfo` both tended to display Han characters rather than the formatted string I was attempting to encode, when I tried to use `utf16` by default when developing `SwiftTaggerID3`.
2) There are only four encoding conventions allowed by the ID3 specs. `utf8`, `isoLatin1`, `utf16`, and `utf16BigEndian`. If the string you attempt to pass contains UNICODE characters that can't be encoded in one of those formats, then `SwiftTaggerID3` will (I think?) simply not encode the frame you're attempting to write. (This is unlikely, but it's possible.) So you will need to make sure any strings you pass to `SwiftTaggerID3` are encoded in one of those four ways.
3) Since I don't have access to the exact data you're attempting to write, I can't be sure this is going to work for you, which is why I'd appreciate it if you'd test it for me.

**Comment by peternguyenxq** [comment]
> Hi Peter!
> 
> Without having the actual data you're working with, I'm only guessing here, but I made some changes to how string encoding was handled, and hopefully it will solve the problem.
> 
> Try using the `encoding-fixes` branch and see if that solves your issues. If it does, I'll go ahead and merge it. If not, I may need to ask my mentor for help, because I'm not all that confident that I'm using the right approach for how determining the string encoding is handled.

Hi @NCrusher74 ,

Thank you for your responses.
This is an example string: **Trên Tình Bạn Dưới Tình Yêu**

I did try your encoding-fixes branch. It worked!
Could you please help to merge to the master branch?

@Coleman and I thank you so much!

**Comment by NCrusher74** [comment]
I will merge that branch within a couple days, but first I want to run something by my mentor, because I want to make sure one of the changes I made isn't going to cause trouble later. I'm glad it worked for you, though!

**Comment by SDGGiesbrecht** [comment]
From what I can see in `encoding-fixes`, nothing there will cause issues _now_. However, some of the unreachable code branches could be vulnerable to becoming reachable if future changes are made elsewhere in the code, in the same manner that the bug surfaced in the first place.

First, to be clear, `String` (or even *a string* in the abstract sense) does not have an inherent encoding. So you are not trying to find _the encoding of the string_, but rather looking for _a viable encoding for the string_. That is trivial, not impossible. I would probably rename the methods accordingly so make it easier to remember and understand in the future. Your way of selecting an encoding—which is correct—does exactly that, by trying to encode the string different ways until it finds one that works.

Of the four encodings you are dealing with—ISO 8859‐1 (a.k.a. Latin‐1), (little endian?) UTF‐16, big endian UTF‐16 and UTF‐8—only the first has a limited character repertoire. The other three can all handle absolutely any Unicode string. So the first `if` that tries ISO 8859‐1 may or may not succeed, but then the second `if` that tries UTF‐16 is guaranteed to succeed, and everything after that is dead code. After trying ISO 8859‐1, I would simply choose one of the UTF variants and select it unconditionally. Choose the one that has widest support among the other applications that will be trying to read the file you are creating.

Once the encoding selection no longer returns an optional, you can clean up a lot of `if`‐`else` code branches at the call sites. Some of those currently do nonsensical things in the dead side of the code branch.

I assume that on the _decoding_ side, you still support all encodings, and still determine the encoding based on what the file declares? Do not attempt any sort of guess‐and‐check with the bytes. _That_ is what your sources were warning you was impossible.

Lastly, I don’t see any new tests. I recommend making a test out of @peternguyenxq’s string, so that you will know immediately if you start to recreate the mistake in the future, and won’t end up releasing a regression.

***

P.S.

If ISO 8859‐1 definitively has wider support among other applications, then it might be worth taking a further step after the bug is fixed and converting all strings to NFC first with Foundation’s [`precomposedStringWithCanonicalMapping`](https://developer.apple.com/documentation/foundation/nsstring/1412645-precomposedstringwithcanonicalma). This could improve interoperability for a small set of strings, because the NFC forms of _„Doppelgänger“_, _« façade »_ and _«jalapeño»_ can be represented in ISO 8859‐1, but not their NFD forms. (Don’t take that as general advice toward NFC though, because NFD is better for most string processing tasks.)

**Comment by NCrusher74** [comment]
Thank you @SDGGiesbrecht !

I will probably default to UTF-8 then, if ISO 8859‐1 doesn't work, because I believe the interoperability problems only arose when attempting to use the UTF-16 encodings.

The order of the four options in the was determined by their order in the ID3 specs for the encoding byte:

```
     $00   ISO-8859-1 [ISO-8859-1]. Terminated with $00.
     $01   UTF-16 [UTF-16] encoded Unicode [UNICODE] with BOM. All
           strings in the same frame SHALL have the same byteorder.
           Terminated with $00 00.
     $02   UTF-16BE [UTF-16] encoded Unicode [UNICODE] without BOM.
           Terminated with $00 00.
     $03   UTF-8 [UTF-8] encoded Unicode [UNICODE]. Terminated with $00.
```

I assumed that the order suggested which encoding was "preferred" by the ID3 spec, hence the idea that ISO-8859-1 should be the default.

But I think the other apps I was checking for interoperability (Yate, MediaInfo, etc) simply didn't like the 16 bit options. I'll double check once I've made the suggested changes and implemented some tests (now that I have a sample string to work with.) But I may be able to simplify things still further and just default to UTF-8 in all cases?

**Comment by SDGGiesbrecht** [comment]
> I believe the interoperability problems only arose when attempting to use the UTF-16 encodings.

> But I may be able to simplify things still further and just default to UTF-8 in all cases?

As long as

- you aren’t encountering interoperability problems with aged but still significant applications, and
- all ID3 versions you want to be able to _write_ support UTF‐8,

then yes, simply using UTF‐8 all the time would be best.

**Comment by SDGGiesbrecht** [comment]
> I assumed that the order suggested which encoding was "preferred" by the ID3 spec

Rather, they are probably in the order that they were added to the specification. Outside of legacy interoperability, older generally isn’t better; otherwise it wouldn’t have needed replacing.

**Comment by NCrusher74** [comment]
> As long as
> 
>     * you aren’t encountering interoperability problems with aged but still significant applications, and
> 
>     * all ID3 versions you want to be able to _write_ support UTF‐8,

Ah, good catch. Turns out UTF-8 isn't supported until version 2.4.

However, in the process of experimenting with interoperability (since it's been over a year since I encountered the issues that led me to conclude that other apps weren't supporting UTF-16, and I wasn't entirely sure if it had just been certain frames that were problematic) none of the other apps are having any problem when I attempt to use UTF-16.

Since I doubt all those apps updated to support UTF-16 in the last year (though I suppose that's possible) I wonder if the notion that they didn't support UTF-16 came from when we were trying to adapt that other library so that we could use it, and the problem was actually in how that other library was encoding strings?

I'm going to run some more tests to be absolutely certain before I decide which way to go with this, but preliminary efforts suggest that UTF-16 isn't as out of the running as I thought it was.
